File: BOJ/2021 ICPC Sinchon Winter Algorithm Camp Contest/20930.cpp
#include <bits/stdc++.h>

#define sz(x) (int)x.size()
#define all(x) x.begin(), x.end()
#define fastio cin.sync_with_stdio(false); cin.tie(nullptr)
using namespace std;
using tt = tuple<int, int, int>;

int find(vector<int> &p, int u) {
    if (p[u] == u) return u;
    return p[u] = find(p, p[u]);
}

bool merge(vector<int> &p, int u, int v) {
    u = find(p, u);
    v = find(p, v);
    if (u == v) return true;
    p[u] = v;
    return false;
}

int main() {
    fastio;
    // input
    int N, Q;
    cin >> N >> Q;
    vector<tt> x_sweep, y_sweep;
    for (int i = 0; i < N; i++) {
        int x0, y0, x1, y1;
        cin >> x0 >> y0 >> x1 >> y1;
        if (x0 > x1) swap(x0, x1);
        if (y0 > y1) swap(y0, y1);
        x_sweep.emplace_back(x0, x1, i);
        y_sweep.emplace_back(y0, y1, i);
    }

    // Sweeping & Union-Find
    vector<int> p(N);
    for (int i = 0; i < N; i++) p[i] = i;
    sort(all(x_sweep));
    sort(all(y_sweep));
    for (auto sweep : {x_sweep, y_sweep}) {
        int before = -1, end_point = -1;
        for (auto [s, e, i] : sweep) {
            if (before < 0 || end_point < s) {
                before = i;
                end_point = e;
            } else {
                merge(p, i, before);
                end_point = max(end_point, e);
            }
        }
    }

    // Query
    while (Q--) {
        int u, v;
        cin >> u >> v;
        cout << int(find(p, u - 1) == find(p, v - 1)) << '\n';
    }
}

File: BOJ/2021 ICPC Sinchon Winter Algorithm Camp Contest/20930.py
from sys import stdin, setrecursionlimit
input = stdin.readline
setrecursionlimit(200001)

def find(p, u):
    if p[u] == u:
        return u
    p[u] = find(p, p[u])
    return p[u]

def union(p, u, v):
    u = find(p, u)
    v = find(p, v)
    if u == v:
        return False
    p[u] = v
    return True

if __name__ == "__main__":
    # input
    N, Q = map(int, input().split())
    x_sweep = []
    y_sweep = []
    for i in range(N):
        x0, y0, x1, y1 = map(int, input().split())
        if x0 > x1:
            x0, x1 = x1, x0
        if y0 > y1:
            y0, y1 = y1, x0
        x_sweep.append((x0, x1, i))
        y_sweep.append((y0, y1, i))

    # Sweeping & Union-Find
    x_sweep.sort()
    y_sweep.sort()
    p = [i for i in range(N)]
    for sweep in [x_sweep, y_sweep]:
        before = -1
        end_point = -1
        for s, e, i in sweep:
            if before < 0 or end_point < s:
                before = i
                end_point = e
            else:
                union(p, i, before)
                end_point = max(end_point, e)

    # Query
    for _ in range(Q):
        u, v = map(int, input().split())
        print(int(find(p, u - 1) == find(p, v - 1)))